docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,157 @@
+```
+The root would be like:
+{
+  "shards" : [
+     {"/": <shard1>},
+     {"/": <shard2>},
+     ...
+   ]
+}
+
+where each shard looks like:
+
+{
+  "blocks" : [
+     {"/": <block1>},
+     {"/": <block2>},
+     ...
+   ]
+}
+
+```
+Each cluster node recursively pins a shard-node of the cluster-dag, ensuring that the blocks of file data referenced underneath the shard are pinned by that node.  The index of shards, i.e. the root, can be (non-recursively) pinned on any/all cluster nodes.
+
+With this implementation the cluster can provide the entire original ipfs file dag on request.  For example if an ipfs peer queries for the entire file they first resolve the root node of the dag which must be pinned on some shard.  If the root has child links then ipfs peers in the cluster are guaranteed to resolve them with content, as all blocks are pinned somewhere in the cluster as children of a shard-node in the cluster-dag.  Note that this implementation conforms well to the goal of coordinating "lightly" above because the cluster state need only keep track of 1 cid per shard.  This prevents the state size from growing too large or more complex than necessary for the purpose of allocation.
+
+The state format would change slightly to account for linking together the root cid of an ipfs file dag and the cluster-dag pinning its leaves
+
+```
+"<cid>": {
+  "name": <pinname>,
+  "cid": <cid>,
+  "clusterdag": <cidcluster>,
+  "allocations" : []
+}
+"<shard1-cid>": {
+  "name": "<pinname>-shard1",
+  "cid": <shard1-cid>,
+  "clusterdag": nil,
+  "allocations": [cluster peers]
+}
+....
+```
+
+Under this implementation replication of shards proceeds like replication of an ordinary pinned cid.  If a shard is replicated across the cluster then data is not always lost during a node failure.  To get better guarantees for less extra space we will need to proceed in the direction of the fault tolerance described in the Future Work section.
+
+
+### Adding files
+Currently the proposed implementation of basic sharding relies on ipfs-cluster having an incoming stream of data that can be read lazily as cluster sends off blocks to remote peers.  This allows sharding files that are too big to fit in a single repo or whose dags are too big to read into memory all at once, as resources can be cleared while streaming occurs.  Only the cids of blocks in the shard need to be maintained to construct the shard node of the cluster-dag, not the entire block or in-memory node structure.  Therefore in our basic sharding implementation we assume that sharded data must be added to ipfs-cluster directly.  This is because, as far as we know, there is not currently a way to access such a stream of blocks belonging to an already-imported-dag without potentially overloading the local ipfs daemon's resources.
```

Comment:
@hsanjuan Are you suggesting we try to incorporate this in our first pass of implementing sharding?  I totally agree this should exist eventually but my understanding was that this is going to be experimental for a while for the reasons listed in this paragraph (please let me know if these reasons are overstated and you think this belongs in our first implementation).  

I could add your suggestion here but would prefer to put it in the later section "Sharding existing dags" which already exists to address this case.  My current view is that this part of the document describes the "basics" that we will work on first and later on I address how we will extend the basics in "Future Work"

Reply:
It's actually easier to do because you don't need the importer part (or your importer is just something that fetches from ipfs). I am wondering if we should do this first. But I also think the importers are really important. In any case, I don't think this is so complex that cannot be done along with the first round of things or shortly after.